### PR TITLE
Apple pay merchant association

### PR DESCRIPTION
**Pull request by danielbowden**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
Our team uses fastlane extensively to automate the entire workflow of 50+ apps. We have been working with Apple Pay and have the need to create Apple Pay Merchant IDs and associate these with the appropriate App IDs for not only these existing applications but all new ones going forward. 

Currently these actions are performed through many steps on the Apple Developer Center:
Creating a new merchant: `Identifiers > Merchant IDs > Add > [name, identifier] > Continue > Register > Done`
Associating with App ID: `Identifiers > App IDs > [App] > Edit > Enable Apple Pay > Edit > [Select Merchant IDs] > Continue > Assign > Done`

I had seen this feature had already been requested and was sitting in an open issue (#9487). It had also come up before in a couple of old issues and I saw that it was yet to be implemented. Late 2016 the ability to enable/disable services such as apple-pay, game-center, in-app-purchase were implemented in PR #6820 / #5100 however there was still the manual step in creating the merchant identifiers and associating them to an app id.

### Description
This PR firstly extends `spaceship` to add the ability to create, list and delete merchants and associate them to app ids.

It exposes this functionality in `produce` mimicking the style of other similar `produce` commands such as `group` and `associate_group`.

I've added tests, JSON stubs and updated the documentation to cover the new changes in both `spaceship` and `produce`.

Examples:

**Creating a new merchant**
```
bundle exec fastlane produce merchant -u xxx@xxx.xxx -o merchant.com.example.app.production -r TestMerchant
[13:38:23]: Starting login with user 'xxx@xxx.xxx'
[13:38:27]: Successfully logged in
[13:38:27]: Creating new merchant 'TestMerchant' with identifier 'merchant.com.example.app.production' on the Apple Dev Center
[13:38:30]: Created merchant SLG6ABT44J
[13:38:30]: Finished creating new merchant 'TestMerchant' on the Dev Center
```

**Associating merchant with app**
```
bundle exec fastlane produce associate_merchant -u xxx@xxx.xxx -a com.example.app merchant.com.example.app.production
[13:41:04]: Starting login with user 'xxx@xxx.xxx'
[13:41:15]: Successfully logged in
[13:41:29]: Validating merchants before association
[13:41:31]: Finalising association with 1 merchant
[13:41:34]: Done!
```

**Associating multiple merchants at once**
```
bundle exec fastlane produce associate_merchant -u xxx@xxx.xxx -a com.example.app merchant.com.example.app.production merchant.com.example.app.sandbox
[13:39:29]: Starting login with user 'xxx@xxx.xxx'
[13:39:33]: Successfully logged in
[13:39:37]: Validating merchants before association
[13:39:40]: Finalising association with 2 merchants
[13:39:43]: Done!
```

### CODE REVIEW

**Comment by danielbowden** [comment]
I signed it!

**Comment by ohayon** [comment]
Hey @danielbowden - thanks for the PR! As this is a big change, this is definitely something we will take a closer look at before merging, so just wanted to give you a heads up that we will likely get to this early next week. 📆 🐙 🚀

**Comment by danielbowden** [comment]
Hi @ohayon, I was just back working on an Apple Pay related app and just came back to follow up on this.

**Comment by ohayon** [comment]
Hey @danielbowden - we have this towards the top of our backlog of PRs to take a closer look at, apologies on the lead time! I am going to try to get to this today!

File: produce/README.md
Comment (ohayon) [inline produce/README.md]: Did you mean to change the indentation on these?

File: produce/README.md
Comment (ohayon) [inline produce/README.md]: Would you mind wrapping these code samples in three backticks rather than relying on indentation?

File: produce/lib/produce/merchant.rb
Comment (ohayon) [inline produce/lib/produce/merchant.rb]: Could we add a check here to only do the `UI.input` if `UI.interactive?` to avoid a potential crash? If `UI.interactive?` returns false, we should raise a `UI.user_error` saying that if the shell is not interactive the `:merchant_identifier` option is required. Does that make sense?

File: produce/lib/produce/merchant.rb
Comment (ohayon) [inline produce/lib/produce/merchant.rb]: Can we wrap the parameter in parentheses?

File: produce/lib/produce/merchant.rb
Comment (ohayon) [inline produce/lib/produce/merchant.rb]: Would you mind adding a comment with some detail around why we would split, capitalize, reverse and then join the ID? Just seems like this line is doing a lot that is a bit unclear. ✨

File: produce/lib/produce/merchant.rb
Comment (ohayon) [inline produce/lib/produce/merchant.rb]: Can we line up the parameter indentation here?

File: produce/lib/produce/merchant.rb
Comment (ohayon) [inline produce/lib/produce/merchant.rb]: Are we sure that this is the only reason why these values wouldn't be equal? Should we maybe change the language to make the suggestion that that might be the issue?

File: produce/lib/produce/merchant.rb
Comment (ohayon) [inline produce/lib/produce/merchant.rb]: Can we reorder these so that the user error comes first (and also wrap `merchant_identifier` as the parameter to `merchant_exists?` in parentheses) and then the success messages, it looks like this will always say that it was created even though the next line says it failed (if I understand this all correctly?)

File: produce/lib/produce/merchant.rb
Comment (ohayon) [inline produce/lib/produce/merchant.rb]: Can we reverse this logic so that we are checking `if app_exists?` and then proceed to the false case in the `else` block?

File: produce/lib/produce/merchant.rb
Comment (ohayon) [inline produce/lib/produce/merchant.rb]: Doesn't this get checked by `app_exists?`. Also `app_exists?` is already making a network call to get the app from spaceship, so it feels like this is a redundant call on line 40?

File: produce/lib/produce/merchant.rb
Comment (ohayon) [inline produce/lib/produce/merchant.rb]: Can we reverse this logic so we can check for the positive case first?

File: produce/lib/produce/merchant.rb
Comment (ohayon) [inline produce/lib/produce/merchant.rb]: Not necessary, but might improve readability... Can we break this into two lines where you check for the platform outside of the arguments?

File: produce/lib/produce/merchant.rb
Comment (ohayon) [inline produce/lib/produce/merchant.rb]: Would it be cool to break out this logic into its own line to avoid doing that check in the string interpolation? 

Also, this could check for the positive case first like `new_merchangs.count == 1 ? merchant : merchants`

File: produce/lib/produce/merchant.rb
Comment (ohayon) [inline produce/lib/produce/merchant.rb]: Isn't this always a string?

File: produce/lib/produce/merchant.rb
Comment (ohayon) [inline produce/lib/produce/merchant.rb]: I'm a bit worried that this get called from a bunch of places and could be causing some redundant network calls? Is it possible to cache this value?

File: produce/lib/produce/merchant.rb
Comment (ohayon) [inline produce/lib/produce/merchant.rb]: Same comment as above with the `merchant_exists?` method.

File: produce/lib/produce/options.rb
Comment (ohayon) [inline produce/lib/produce/options.rb]: Can we put in the description that the value should start with `merchant.` or, maybe we can be smart about it and prepend to the string if it doesn't already start with that string?

File: produce/lib/produce/options.rb
Comment (ohayon) [inline produce/lib/produce/options.rb]: Can we fix up the indentation here?

**Review by ohayon — CHANGES_REQUESTED** [review]
Hey @danielbowden - Thanks again for this PR it is definitely really awesome and I would love to get it merged in! I left a bunch of comments and am curious to hear if you have any thoughts about them. For the most part they are style things and this is 💯 , let me know if you have any questions about any of my comments. 🚀

File: produce/lib/produce/merchant.rb
Comment (KrauseFx) [inline produce/lib/produce/merchant.rb]: Is this PR something people usually run on CI? If not, we might just leave it with a crash to not over-engineer?

File: produce/README.md
Comment (KrauseFx) [inline produce/README.md]: Yes, nice catch, the same applies for all newly added lines below also

File: produce/README.md
Comment (KrauseFx) [inline produce/README.md]: > Use `--help` for more information about all available parameters

File: produce/lib/produce/merchant.rb
Comment (KrauseFx) [inline produce/lib/produce/merchant.rb]: I wish there was a rubocop rule for this...

File: produce/lib/produce/merchant.rb
Comment (KrauseFx) [inline produce/lib/produce/merchant.rb]: I thought there is a rubocop rule for this, but maybe we disabled it, not a big deal though.

File: produce/lib/produce/merchant.rb
Comment (KrauseFx) [inline produce/lib/produce/merchant.rb]: Nice idea

File: produce/lib/produce/options.rb
Comment (KrauseFx) [inline produce/lib/produce/options.rb]: I don't have the full context here, but from my understanding this would also show up when running `fastlane produce --help`, right? If so, what happens if the user provides this for the regular produce command, and not the one for merchants? You mention the other `--help` command in the README, doesn't this show the same options than the main one? Just asking some questions here, I haven't checked myself.

File: spaceship/docs/DeveloperPortal.md
Comment (KrauseFx) [inline spaceship/docs/DeveloperPortal.md]: Can you fix the spacing of the second parameter here

File: produce/lib/produce/merchant.rb
Comment (danielbowden) [inline produce/lib/produce/merchant.rb]: @ohayon very good question, I'm not sure? I copied that from here: https://github.com/fastlane/fastlane/blob/master/produce/lib/produce/group.rb#L78

**Comment by danielbowden** [comment]
Thanks for reviewing @ohayon and @KrauseFx, much appreciated and lots of good catches in there. I've made some changes and style fixes, I'll comment inline on your reviews. 

I agree with your comments, I had actually tried to copy much of the same style, structure and methods from existing code in produce such as `group.rb` and the `associate_group` command to keep it consistent.

File: produce/README.md
Comment (danielbowden) [inline produce/README.md]: Sorry no, fixed.

File: produce/README.md
Comment (danielbowden) [inline produce/README.md]: Agreed, changed. I was copying the existing style of the code samples above for `Enabling / Disabling Application Services`

File: produce/lib/produce/merchant.rb
Comment (danielbowden) [inline produce/lib/produce/merchant.rb]: Haha fixed. I was copying the style in group.rb [here](https://github.com/fastlane/fastlane/blob/master/produce/lib/produce/group.rb#L12) but agree the parentheses is much more readble

File: produce/lib/produce/merchant.rb
Comment (danielbowden) [inline produce/lib/produce/merchant.rb]: @ohayon This was already used in https://github.com/fastlane/fastlane/blob/master/produce/lib/produce/group.rb#L17 as a way of creating a sensible default name from the provided identifier so I copied it here. 

ie. merchant.com.example.app.sandbox is given a name of "Sandbox App Example Com Merchant" if no specific merchant_name was provided

 I'm happy to change it or add a comment in code to give more detail if you like?

File: produce/lib/produce/merchant.rb
Comment (danielbowden) [inline produce/lib/produce/merchant.rb]: Sorry I'm not sure, this is the same as https://github.com/fastlane/fastlane/blob/master/produce/lib/produce/group.rb#L25

I have changed the language slightly in my latest commits to instead read "Your merchant name **might** include..."

File: produce/lib/produce/merchant.rb
Comment (danielbowden) [inline produce/lib/produce/merchant.rb]: Done 👍

File: produce/lib/produce/merchant.rb
Comment (danielbowden) [inline produce/lib/produce/merchant.rb]: Done 👍

File: produce/lib/produce/merchant.rb
Comment (danielbowden) [inline produce/lib/produce/merchant.rb]: @ohayon Good catch and nice idea. I've cached the value in my latest commits.

File: produce/lib/produce/options.rb
Comment (danielbowden) [inline produce/lib/produce/options.rb]: @KrauseFx good catch! I have fixed this up. I couldn't find a good existing example of passing the extra options when they aren't related to the regular command so can you please review my changes in `produce/commands_generator.rb`

File: produce/lib/produce/options.rb
Comment (danielbowden) [inline produce/lib/produce/options.rb]: @ohayon I have removed this verify based on @KrauseFx's related comment below. My latest change now prepends 'merchant' to the identifier string if it doesn't already start with it as per your suggestion. 👍

**Comment by danielbowden** [comment]
@ohayon @KrauseFx Sorry for the spam. I've just made comments inline to your comments to help with context 👍

File: produce/lib/produce/merchant.rb
Comment (ohayon) [inline produce/lib/produce/merchant.rb]: I think it should always be a string 😄 . Lets remove the `to_s` for now.

File: produce/lib/produce/options.rb
Comment (ohayon) [inline produce/lib/produce/options.rb]: How come these are removed?

File: produce/lib/produce/commands_generator.rb
Comment (ohayon) [inline produce/lib/produce/commands_generator.rb]: I saw that you removed these from the options file, just curious to know the reason for that? I think its always helpful to have all available options in the `<tool name>/options.rb` file as a sort of documentation.

File: produce/lib/produce/merchant.rb
Comment (ohayon) [inline produce/lib/produce/merchant.rb]: Ah, thanks for citing that other file with the same pattern, yeah I think if we could break it out into more digestible lines that would be much better, we should probably also do the same in the file you referenced too 😄

File: produce/lib/produce/merchant.rb
Comment (ohayon) [inline produce/lib/produce/merchant.rb]: ❤️

File: produce/lib/produce/merchant.rb
Comment (ohayon) [inline produce/lib/produce/merchant.rb]: This would be a great method to have in a monkey patch on Array in `fastlane_core` don't worry about making any changes to it now, just a thought that we use this type of logic all over and it would be nice to have a central place for it! 💯

File: produce/lib/produce/merchant.rb
Comment (ohayon) [inline produce/lib/produce/merchant.rb]: I see whats going on here, but could you maybe just add a comment for new comers into this code?

File: produce/lib/produce/merchant.rb
Comment (ohayon) [inline produce/lib/produce/merchant.rb]: Sorry if I am missing something here - is there a reason why we have this as a class level method and an instance method?

File: produce/lib/produce/merchant.rb
Comment (ohayon) [inline produce/lib/produce/merchant.rb]: Same thing as above regarding two levels of the same method?

**Review by ohayon — CHANGES_REQUESTED** [review]
Hey @danielbowden! Thanks so much for making all those changes! I just left a couple more comments and we should be able to get this over the finish line soon!

File: spaceship/lib/spaceship/portal/merchant.rb
Comment (snatchev) [inline spaceship/lib/spaceship/portal/merchant.rb]: Please remove usage of `factory`. It has been copied into too many places (no judgement, I recognize it's natural to reference existing implementations).

The usage of `factory` should be limited to very specific use-cases described here: https://github.com/fastlane/fastlane/blob/3ffaaf2d0befc2df3132b7c266bd775aa13f49b7/spaceship/lib/spaceship/base.rb#L177

File: spaceship/spec/portal/portal_stubbing.rb
Comment (snatchev) [inline spaceship/spec/portal/portal_stubbing.rb]: I would love to move away from this style of request stubbing. For an example of the new style, please check out `key_spec.rb` and `portal_client_spec.rb` in https://github.com/fastlane/fastlane/pull/9560/files#diff-427c42b9c7ea4456fc2eee77cf6e4b1b 

The high-level goal with this change is to make the fixtures more flexible by putting them in code rather then brittle fixture files. Also, we now separate the tests into two different concerns. First is making the expected HTTP request ( as in `portal_client_spec.rb`) and the other ensures that the data model (`key_spec.rb`) behaves as expected based on received data from the API.

**Review by snatchev — CHANGES_REQUESTED** [review]
Thank you for adding these changes to spaceship! Overall, it's pretty good, but I would love to see the tests in the style of this PR https://github.com/fastlane/fastlane/pull/9560/files

Also you can reference specs like:
https://github.com/fastlane/fastlane/blob/1ad43dd4a101c7521d0bbf7c012d600f8f708e77/spaceship/spec/test_flight/build_trains_spec.rb - data model spec

https://github.com/fastlane/fastlane/blob/1ad43dd4a101c7521d0bbf7c012d600f8f708e77/spaceship/spec/test_flight/client_spec.rb - client spec

File: produce/lib/produce/options.rb
Comment (danielbowden) [inline produce/lib/produce/options.rb]: @ohayon These are removed following @KrauseFx's comment above about the new options appearing when they are not relevant when using `bundle exec produce --help`. They are not options for the other commands such as `create` and `group` and should not be included. With the latest changes this no longer happens yet they will still appear in help for the specific merchant command, eg `bundle exec fastlane produce merchant --help`

File: produce/lib/produce/commands_generator.rb
Comment (danielbowden) [inline produce/lib/produce/commands_generator.rb]: @ohayon I've just answered in more detail in your other review comment on `options.rb`. It was removed so that the options do not show for unrelated commands, yet will still show for the merchant command

File: produce/lib/produce/merchant.rb
Comment (danielbowden) [inline produce/lib/produce/merchant.rb]: @ohayon I've split the merchant name out into a separate method and broken it apart as well as added a test into the new `merchant_spec.rb` in my latest commit.

File: produce/lib/produce/merchant.rb
Comment (danielbowden) [inline produce/lib/produce/merchant.rb]: No worries, done 👍

File: produce/lib/produce/merchant.rb
Comment (danielbowden) [inline produce/lib/produce/merchant.rb]: I'm sorry, in a rush to get this out last Friday afternoon I didn't add a `merchant_spec.rb`. I've added this in my latest commit. This class level method is to simplify testing by not having to instantiate the merchant object and also to not have to mock unrelated methods which would make them more brittle to refactoring.

File: produce/lib/produce/merchant.rb
Comment (danielbowden) [inline produce/lib/produce/merchant.rb]: @ohayon I'm sorry, in a rush to get this out last Friday afternoon I didn't add a `merchant_spec.rb`. I've added this in my latest commit. This class level method is to simplify testing by not having to instantiate the merchant object and also to not have to mock unrelated methods which would make them more brittle to refactoring.

File: spaceship/lib/spaceship/portal/merchant.rb
Comment (danielbowden) [inline spaceship/lib/spaceship/portal/merchant.rb]: Done 👍

**Comment by danielbowden** [comment]
@ohayon I've made the requested changes and also addressed your comments inline above.

**Comment by danielbowden** [comment]
@snatchev I have just added some changes to move to the newer request stubbing style you highlighted, replacing fixture file stubbing with new in-code request stubbing style.

**Review by snatchev — APPROVED** [review]

**Comment by snatchev** [comment]
@danielbowden thank you! The changes look great.

**Comment by danielbowden** [comment]
Great to hear, thanks @snatchev

**Comment by KrauseFx** [comment]
This is awesome, can't wait to have this merged and shipped 🚀  Thanks for your work @danielbowden 👍  Can you just resolve the remaining merge conflicts? :)

**Comment by danielbowden** [comment]
@KrauseFx Conflicts now resolved 👍

**Comment by KrauseFx** [comment]
Alright, thanks, this looks good to me cc @ohayon

**Review by ohayon — APPROVED** [review]

**Comment by ohayon** [comment]
Thanks @danielbowden !! 🚀

**Comment by EricNetsch** [comment]
@danielbowden  Really fantastic job here. This is exactly what we needed. How can I implement this into my Fastfile? 
Sorry if this is a noob question, just not clear in the current docs. 
Thanks!

File: produce/README.md
Comment (EricNetsch) [inline produce/README.md]: Can we please add these examples to the Fastfile example bellow if possible. I am a bit confused on how to implement the new merchant params. Thanks!

**Comment by danielbowden** [comment]
Thanks @EricKuck! Glad someone else finds it useful, exactly why I added it as our team needed it too.

Not a noob question, I will need to do the same thing for our Fastfile too. The secondary commands for produce (and other fastlane tools) are by default only available through CLI. Some already have [actions](https://docs.fastlane.tools/actions/) available to expose this spaceship/tool functionality, eg [resign for sigh](https://docs.fastlane.tools/actions/#resign).

The quick shortcut way of doing this in your Fastfile is running it as a shell command:
```ruby
sh "bundle exec fastlane produce merchant --merchant_name MyMerchant --merchant_identifier merchant.com.appname.mymerchant"
```

The cleaner way is to create a [local action](https://docs.fastlane.tools/plugins/create-plugin/#local-actions). I have just had a quick try at creating an action by calling the Spaceship classes directly. This doesn't give you the added benefits of calling it via Produce, benefits such as appending 'merchant' to the front of your identifier if you left it off, checking the merchant exists before trying to create it, etc. So perhaps @KrauseFx can suggest the proper way of doing this?

```ruby
class CreateMerchantAction < Action
      def self.run(params)
        require 'spaceship'
        credentials = CredentialsManager::AccountManager.new(user: params[:username])
        Spaceship.login(credentials.user, credentials.password)
        Spaceship.select_team
        Spaceship.merchant.create!(bundle_id: params[:merchant_identifier], name: params[:merchant_name], mac: false)
end

...

def self.available_options
        user = CredentialsManager::AppfileConfig.try_fetch_value(:apple_dev_portal_id)
        user ||= CredentialsManager::AppfileConfig.try_fetch_value(:apple_id)

        [
          FastlaneCore::ConfigItem.new(key: :merchant_name,
                                       env_name: "FL_CREATE_MERCHANT_NAME", # The name of the environment variable
                                       description: "Name for the merchant that is created" # a short description of this parameter
                                       ),
          FastlaneCore::ConfigItem.new(key: :merchant_identifier,
                                       env_name: "FL_CREATE_MERCHANT_IDENTIFIER",
                                       description: "Merchant identifier for the merchant"
                                       ),
          FastlaneCore::ConfigItem.new(key: :username,
                                       env_name: "FL_CREATE_MERCHANT_USERNAME",
                                       description: "Optional: Your Apple ID",
                                       default_value: user)
        ]
end
```

This could stay as a local action in your own repository. I know actions are not as accepted into the main repo these days in favour of plugins but maybe as it is more like core developer center functionality that bridges the existing code it could get added to the main repo. cc @KrauseFx

**Comment by danielbowden** [comment]
@EricNetsch I've just taken another run at this now that we need to use it in our workflow. I have created a [local action](https://docs.fastlane.tools/plugins/create-plugin/#local-actions) as mentioned before but now have got it working using the correct Produce/Merchant classes instead. So now it does include the additional benefits I had mentioned in my last comment. I've put it up as a public gist:

https://gist.github.com/danielbowden/49ee4cbd06d969fca5894ed89a21f17d

**Comment by KrauseFx** [comment]
Oh wow, that's some great content right here :) Do you want to add this to the docs so we have it in one central space? :)

**Comment by danielbowden** [comment]
Sorry @KrauseFx which bit did you mean? I'm happy to add to the docs.

Also, would you like a PR with those two actions I linked in my gists that provides access to the merchant functionality in produce? I know the team doesn't really take on more action PRs but this one is to core functionality not a third party service.

**Comment by EricNetsch** [comment]
@danielbowden 👏 great work ! Going to test this in our workflow today. I will let you know how everything goes. Thanks!

**Pull request merged**
